libraries/lib-module-manager/AsyncPluginValidator.cpp

```diff
@@ -0,0 +1,237 @@
+   {
+      try
+      {
+         BasicUI::CallAfter([this, self = weak_from_this(), msg]
+         {
+            if(auto lck = self.lock() && mDelegate != nullptr)
+               mDelegate->OnInternalError(msg);
+         });
+      }
+      catch(...)
+      {
+         //no way to report about a problem, though shouldn't be the case...
+      }
+   }
+
+   void HandleResult(details::PluginValidationResult&& result) noexcept
+   {
+      try
+      {
+         BasicUI::CallAfter([this, self = weak_from_this(), result = details::PluginValidationResult { result }]
+         {
+            if(auto lck_self = self.lock())
+            {
+               if(mDelegate == nullptr)
+                  return;
+
+               //Release the current value to keep state invariant
+               //Caller is free to use Validate now
+               std::optional<wxString> request;
+               {
+                  std::lock_guard lck_sync(mSync);
+                  mRequest.swap(request);
+               }
+
+               if(!request.has_value())
+               {
+                  //Invalid state error
+                  mDelegate->OnInternalError(result.GetErrorMessage());
+                  return;
+               }
+
+               if(result.IsValid())
+               {
+                  for(auto& desc : result.GetDescriptors())
+                     mDelegate->OnPluginFound(PluginDescriptor { desc });
+               }
+               else
+               {
+                  wxString providerId;
+                  wxString pluginPath;
+                  details::ParseRequestString(*request, providerId, pluginPath);
+
+                  //Even if plugin validation has failed or no effects were discovered,
+                  //create an entry for it but make it disabled by default
+                  PluginID ID = providerId + wxT("_") + pluginPath;
+                  PluginDescriptor pluginDescriptor;
+                  pluginDescriptor.SetPluginType(PluginTypeStub);
+                  pluginDescriptor.SetID(ID);
+                  pluginDescriptor.SetProviderID(providerId);
+                  pluginDescriptor.SetPath(pluginPath);
+                  pluginDescriptor.SetEnabled(false);
+                  pluginDescriptor.SetValid(false);
+
+                  mDelegate->OnPluginFound(pluginDescriptor);
+               }
+               mDelegate->OnValidationFinished();
+            }
+         });
+      }
+      catch(...)
+      {
+         ///no way to report about a problem, though shouldn't be the case...
+      }
+   }
+
+public:
+
+   Impl(Impl&) = delete;
+   Impl(Impl&&) = delete;
+   Impl& operator=(Impl&) = delete;
+   Impl& operator=(Impl&&) = delete;
+
+   Impl(Delegate& delegate) : mDelegate(&delegate) { }
+
+   ~Impl() override
+   {
+      //important to reset delegate before IPCChannelStatusCallback::OnDisconnect
+      //is called by server
+      mDelegate = nullptr;
```

Comment:
A thread join happens at the closing brace, so callbacks to members may happen before that completes, making mDelegate null matters while joining.  Am I right?  It may help to explain a little more.

Reply:
Yes, it's correct